### PR TITLE
New resource: `data azurerm_locations`

### PR DESCRIPTION
**Pull request by HappyTobi**
Hi all,

I created a new "data" resource called `azurerm_locations` to close the issue #3025

Regards
Tobi

### CODE REVIEW

**Comment by HappyTobi** [comment]
Hi all,

@tombuildsstuff do you think you can take a look?

Thank you
Regards

**Comment by HappyTobi** [comment]
Hi @tombuildsstuff do you have a chance to review my pr?

**Comment by ms-zhenhua** [comment]
Hi @HappyTobi, thank you for contributing this PR. According to the latest [Contributor Guide](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/topics/best-practices.md#typed-vs-untyped-resources), the provider accepts `Typed` resources only. Could you kindly update your code to a `Typed` resource, then we can have another look. You may use [this PR](https://github.com/hashicorp/terraform-provider-azurerm/pull/22277/files) as an example.

**Comment by HappyTobi** [comment]
@ms-zhenhua can you please review the implementation again.

Test output:
```bash
make acctests SERVICE='subscription' TESTARGS='-run=TestAccDataSourceSubscription_current' TESTTIMEOUT='60m' 
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./internal/services/subscription -run=TestAccDataSourceSubscription_current -timeout 60m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccDataSourceSubscription_current
=== PAUSE TestAccDataSourceSubscription_current
=== CONT  TestAccDataSourceSubscription_current
--- PASS: TestAccDataSourceSubscription_current (13.82s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/subscription  13.845s
```

**Comment by ms-zhenhua** [comment]
> @ms-zhenhua can you please review the implementation again.
> 
> Test output:
> 
> ```shell
> make acctests SERVICE='subscription' TESTARGS='-run=TestAccDataSourceSubscription_current' TESTTIMEOUT='60m' 
> ==> Checking that code complies with gofmt requirements...
> ==> Checking that Custom Timeouts are used...
> ==> Checking that acceptance test packages are used...
> TF_ACC=1 go test -v ./internal/services/subscription -run=TestAccDataSourceSubscription_current -timeout 60m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
> === RUN   TestAccDataSourceSubscription_current
> === PAUSE TestAccDataSourceSubscription_current
> === CONT  TestAccDataSourceSubscription_current
> --- PASS: TestAccDataSourceSubscription_current (13.82s)
> PASS
> ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/subscription  13.845s
> ```

Hi @HappyTobi, thank you for your updates. But `TestAccDataSourceSubscription_current` is not the testcases which you added this time. Could you provide the test results of the testcases added in this PR?
BTW, since Terraform natively supports [filtering](https://developer.hashicorp.com/terraform/language/expressions/for#filtering-elements), do we still need to add the `filter` property?

**Comment by HappyTobi** [comment]
Hi @ms-zhenhua 
Thanks for the feedback.
I removed the filter, I think simpler and less code is much better than a "data source" filter.

Here is also the test output after the code was  updated:

```bash
make acctests SERVICE='subscription' TESTARGS='-run=TestAccDataSourceLocations_westUS' TESTTIMEOUT='60m'
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./internal/services/subscription -run=TestAccDataSourceLocations_westUS -timeout 60m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccDataSourceLocations_westUS
=== PAUSE TestAccDataSourceLocations_westUS
=== CONT  TestAccDataSourceLocations_westUS
--- PASS: TestAccDataSourceLocations_westUS (30.64s)
PASS
ok  	github.com/hashicorp/terraform-provider-azurerm/internal/services/subscription	32.379s
```

other test:
```bash
make acctests SERVICE='subscription' TESTARGS='-run=TestAccDataSourceLocations_NonExistingRegion' TESTTIMEOUT='60m'
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./internal/services/subscription -run=TestAccDataSourceLocations_NonExistingRegion -timeout 60m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccDataSourceLocations_NonExistingRegion
=== PAUSE TestAccDataSourceLocations_NonExistingRegion
=== CONT  TestAccDataSourceLocations_NonExistingRegion
--- PASS: TestAccDataSourceLocations_NonExistingRegion (15.77s)
PASS
ok  	github.com/hashicorp/terraform-provider-azurerm/internal/services/subscription	17.490s
```

File: internal/services/subscription/data_source_locations.go [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+	if err != nil {
+		return err
+	}
+
+	state := getZonesMapping(locationValue)
+	state.RegionType = string(*locationValue.Metadata.RegionType)
+	state.DisplayName = *locationValue.DisplayName
+	state.Location = normalizedLocation
+
+	metadata.SetID(subscriptionLocation)
```
Comment (ms-zhenhua) [inline internal/services/subscription/data_source_locations.go:141]:
```suggestion
	metadata.SetID(fmt.Sprintf("%s/locations/%s", id.ID(), normalizedLocation))
```

File: internal/services/subscription/data_source_locations.go [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+		return fmt.Errorf("retrieving %s: %+v", id, err)
+	}
+
+	var model ManagerSubscriptionLocationsDataSourceModel
+	if err := metadata.Decode(&model); err != nil {
+		return fmt.Errorf("decoding: %+v", err)
+	}
+
+	normalizedLocation := location.Normalize(model.Location)
+	subscriptionLocation := parse.NewSubscriptionLocation(id.SubscriptionId, normalizedLocation)
```
Comment (ms-zhenhua) [inline internal/services/subscription/data_source_locations.go:129]: maybe the `parse` package can be removed if it's only used for setting the ID?

File: internal/services/subscription/data_source_locations.go [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+	ZoneMappings []SubscriptionLocationsZoneMapping `tfschema:"zone_mappings"`
+}
+
+type SubscriptionLocationsZoneMapping struct {
+	LogicalZone  string `tfschema:"logical_zone"`
+	PhysicalZone string `tfschema:"physical_zone"`
+}
+
+func (r ManagerSubscriptionLocationsDataSource) ResourceType() string {
+	return "azurerm_locations"
```
Comment (ms-zhenhua) [inline internal/services/subscription/data_source_locations.go:43]: maybe it should be named `azurerm_location` as it only gets one specific location data by `location`? And all `locations` should be replaced with `location`?

File: internal/services/subscription/data_source_locations.go [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+
+	var model ManagerSubscriptionLocationsDataSourceModel
+	if err := metadata.Decode(&model); err != nil {
+		return fmt.Errorf("decoding: %+v", err)
+	}
+
+	normalizedLocation := location.Normalize(model.Location)
+	subscriptionLocation := parse.NewSubscriptionLocation(id.SubscriptionId, normalizedLocation)
+
+	locationValue, err := filterLocation(normalizedLocation, resp.Model.Value)
```
Comment (ms-zhenhua) [inline internal/services/subscription/data_source_locations.go:131]:
```suggestion
	locationValue, err := getLocation(normalizedLocation, resp.Model.Value)
```

File: internal/services/subscription/data_source_locations.go [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+	if err != nil {
+		return err
+	}
+
+	state := getZonesMapping(locationValue)
```
Comment (ms-zhenhua) [inline internal/services/subscription/data_source_locations.go:136]:
```suggestion
	state.ZoneMappings := flattenZonesMapping(locationValue)
```
Shall we define `var state LocationDataSourceModel` at the top level like other existing data sources?

File: internal/services/subscription/data_source_locations.go [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+	state.RegionType = string(*locationValue.Metadata.RegionType)
```
Comment (ms-zhenhua) [inline internal/services/subscription/data_source_locations.go:137]:
```suggestion
	state.RegionType = string(pointer.From(locationValue.Metadata.RegionType))
```

File: internal/services/subscription/data_source_locations.go [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+	state.DisplayName = *locationValue.DisplayName
```
Comment (ms-zhenhua) [inline internal/services/subscription/data_source_locations.go:138]:
```suggestion
	state.DisplayName = pointer.From(locationValue.DisplayName)
```

File: internal/services/subscription/data_source_locations.go [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+	state.Location = normalizedLocation
+
+	metadata.SetID(subscriptionLocation)
+
+	return metadata.Encode(&state)
+}
+
+func filterLocation(location string, input *[]resourcesSubscription.Location) (*resourcesSubscription.Location, error) {
```
Comment (ms-zhenhua) [inline internal/services/subscription/data_source_locations.go:146]:
```suggestion
func getLocation(location string, input *[]resourcesSubscription.Location) (*resourcesSubscription.Location, error) {
```

File: internal/services/subscription/data_source_locations.go [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+	for _, item := range *input {
```
Comment (ms-zhenhua) [inline internal/services/subscription/data_source_locations.go:147]: what if `input == nil` ?

File: internal/services/subscription/data_source_locations.go [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+}
+
+func getZonesMapping(location *resourcesSubscription.Location) ManagerSubscriptionLocationsDataSourceModel {
+	state := ManagerSubscriptionLocationsDataSourceModel{
+		ZoneIds:      make([]string, 0),
+		ZoneNames:    make([]string, 0),
+		ZoneMappings: make([]SubscriptionLocationsZoneMapping, 0),
+	}
+
+	for _, zoneMapping := range *location.AvailabilityZoneMappings {
```
Comment (ms-zhenhua) [inline internal/services/subscription/data_source_locations.go:163]: what if `location.AvailabilityZoneMappings == nil` ?

File: website/docs/d/locations.html.markdown [hunk 11]
```diff
@@ -0,0 +1,67 @@
+---
+subcategory: "Base"
+layout: "azurerm"
+page_title: "Azure Resource Manager: azurerm_locations"
+description: |-
+  This data source return the available Locations for a specific Azure Region.
```
Comment (ms-zhenhua) [inline website/docs/d/locations.html.markdown:6]:
```suggestion
  Get information of a specific location.
```

File: website/docs/d/locations.html.markdown [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+---
+
+# Data Source: azurerm_ocations
+
+This data source return the available Locations for a specific Azure Region.
```
Comment (ms-zhenhua) [inline website/docs/d/locations.html.markdown:11]:
```suggestion
Use this data source to access information of a specific location.
```

File: website/docs/d/locations.html.markdown [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+
+## Example Usage
+
+```hcl
+data "azurerm_locations" "example" {
+  location = "West Europe"
+
+  filter {
```
Comment (ms-zhenhua) [inline website/docs/d/locations.html.markdown:19]: `filter` should be removed ?

File: website/docs/d/locations.html.markdown [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+    exclude_zone_ids = ["1"]
+  }
+}
+```
+
+## Argument Reference
+
+* `location` - (Required) The Azure location to retrieve the Extended Locations for.
```
Comment (ms-zhenhua) [inline website/docs/d/locations.html.markdown:27]:
```suggestion
* `location` - (Required) Specifies the supported Azure location where the resource exists.
```

File: website/docs/d/locations.html.markdown [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+
+## Attributes Reference
+
+* `id` - The ID of Location within this Subscription.
+
+* `region_type` - The available region type `Physical`, `Logical`
```
Comment (ms-zhenhua) [inline website/docs/d/locations.html.markdown:33]:
```suggestion
* `region_type` - The available region type. Possible values are `Physical` and `Logical`.
```

File: website/docs/d/locations.html.markdown [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+
+* `display_name` - The display name of the location.
+
+* `zone_ids` - The logical zone id's for the availability zone
+
+* `zone_names` - The fully qualified physical zone name of availability zone
+
+* `zone_mappings` - The available zone mappings for the region
```
Comment (ms-zhenhua) [inline website/docs/d/locations.html.markdown:41]:
```suggestion
* `zone_mappings` - A `zone_mappings` block as defined below.
```

File: website/docs/d/locations.html.markdown [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+
+* `filter` - (Optional) Filter
+
+---
+
+The `zone_mappings` block supports the following:
```
Comment (ms-zhenhua) [inline website/docs/d/locations.html.markdown:47]:
```suggestion
A `zone_mappings` block exports the following:
```

**Review by ms-zhenhua — CHANGES_REQUESTED** [review]
Hi @HappyTobi , I've taken a look through and left some comments inline. Kindly confirm. Once that's done I can take another look over. Thanks.

File: internal/services/subscription/data_source_locations.go [hunk 18]
```diff
@@ -0,0 +1,175 @@
...
+	return map[string]*pluginsdk.Schema{
+		"region_type": {
+			Type:     pluginsdk.TypeString,
+			Computed: true,
+		},
+		"display_name": {
+			Type:     pluginsdk.TypeString,
+			Computed: true,
+		},
+		"zone_ids": {
```
[the remark this answers is not in the document]
Comment (HappyTobi) [reply internal/services/subscription/data_source_locations.go:69]: IMHO the good thing about keeping them separat is, that you can use for example the zone_ids directly at other resources.

File: internal/services/subscription/data_source_locations.go (hunk 3 above)
Comment (HappyTobi) [reply internal/services/subscription/data_source_locations.go:43]: fine for me

File: internal/services/subscription/data_source_locations.go (hunk 9 above)
Comment (HappyTobi) [reply internal/services/subscription/data_source_locations.go:147]: Added a check to the azure client response

File: internal/services/subscription/data_source_locations.go (hunk 10 above)
Comment (HappyTobi) [reply internal/services/subscription/data_source_locations.go:163]: Added a check to the azure client response, else the code flow will exit early

**Comment by HappyTobi** [comment]
Hi @ms-zhenhua 
I updated the implementation and run the tests again:


```bash
make acctests SERVICE='subscription' TESTARGS='-run=TestAccDataSourceLocations_westUS' TESTTIMEOUT='60m'
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./internal/services/subscription -run=TestAccDataSourceLocations_westUS -timeout 60m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccDataSourceLocations_westUS
=== PAUSE TestAccDataSourceLocations_westUS
=== CONT  TestAccDataSourceLocations_westUS
--- PASS: TestAccDataSourceLocations_westUS (29.20s)
PASS
```

```bash
make acctests SERVICE='subscription' TESTARGS='-run=TestAccDataSourceLocations_westUS' TESTTIMEOUT='60m'
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./internal/services/subscription -run=TestAccDataSourceLocations_westUS -timeout 60m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccDataSourceLocations_westUS
=== PAUSE TestAccDataSourceLocations_westUS
=== CONT  TestAccDataSourceLocations_westUS
--- PASS: TestAccDataSourceLocations_westUS (29.20s)
PASS
```

File: internal/services/subscription/data_source_locations.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (ms-zhenhua) [reply internal/services/subscription/data_source_locations.go:63]: Terraform natively supports [expressions ](https://developer.hashicorp.com/terraform/language/expressions/for)to complete this kind of work. For example, we can use `[for item in data.azurerm_location.zone_mappings : item.logical_zone]` to get the zone_ids.

File: internal/services/subscription/data_source_locations.go (hunk 10 above)
Comment (ms-zhenhua) [reply internal/services/subscription/data_source_locations.go:143]: I do not see the check of `location.AvailabilityZoneMappings` outside this method. It'd be better directly check it in this method like [other resources](https://github.com/hashicorp/terraform-provider-azurerm/blob/e3c013b04cbaf825939439fb576a5479608683a5/internal/services/subscription/extended_locations_data_source.go#L75).

**Review by ms-zhenhua — CHANGES_REQUESTED** [review]
Thanks for your update @HappyTobi. Some minor comments otherwise this looks fine.

File: internal/services/subscription/data_source_location.go [hunk 19]
```diff
@@ -0,0 +1,157 @@
...
+		}
+	}
+
+	return nil, fmt.Errorf("no location was found for %q", location)
+}
+
+func flattenZonesMapping(location *resourcesSubscription.Location) (zoneMappings []SubscriptionLocationsZoneMapping) {
+	zoneMappings = make([]SubscriptionLocationsZoneMapping, 0)
+
+	if location == nil {
```
Comment (ms-zhenhua) [inline internal/services/subscription/data_source_location.go:143]: I think we should also do the `nil` check for `location.AvailabilityZoneMappings`?

Comment (ms-zhenhua) [inline internal/services/subscription/data_source_location.go:143]: I think we should also do the `nil` check for `location.AvailabilityZoneMappings`?

**Review by ms-zhenhua — CHANGES_REQUESTED** [review]

File: internal/services/subscription/data_source_location.go (hunk 19 above)
Comment (HappyTobi) [reply internal/services/subscription/data_source_location.go]: I added the check

**Review by ms-zhenhua — CHANGES_REQUESTED** [review]
@HappyTobi Thank you for your updates. LGTM except some minor comments on naming. Please also fix `branch conflicts` issue.

image

**Comment by ms-zhenhua** [comment]
@HappyTobi Thank you for your updates. Could you kindly confirm the following testcase error?

image

**Comment by HappyTobi** [comment]
> @HappyTobi Thank you for your updates. Could you kindly confirm the following testcase error?
> 
> image

I run that test multiple times and didn't see that issue.
I will update the test to just check that the "physical_zone" is not empty.

**Comment by ms-zhenhua** [comment]
@HappyTobi , thanks for your updates. LGTM. Please update the issue title to `azurerm_location`

image

File: internal/services/subscription/data_source_location.go
Comment (catriona-m) [inline internal/services/subscription/data_source_location.go:1]: can we rename this to `location_data_source.go`?

File: internal/services/subscription/data_source_location.go [hunk 20]
```diff
@@ -0,0 +1,157 @@
...
+		"location": commonschema.LocationWithoutForceNew(),
+	}
+}
+
+func (r LocationDataSource) Attributes() map[string]*pluginsdk.Schema {
+	return map[string]*pluginsdk.Schema{
+		"region_type": {
+			Type:     pluginsdk.TypeString,
+			Computed: true,
+		},
```
Comment (catriona-m) [inline internal/services/subscription/data_source_location.go:56]: within the azurerm provider we use `location` to mean a physical location (which supports extended locations and custom locations) - given this, we should remove this field and filter only to the physical location in the Read() function

File: internal/services/subscription/data_source_location_test.go
Comment (catriona-m) [inline internal/services/subscription/data_source_location_test.go:1]: can we rename this to `location_data_source_test.go`?

File: internal/services/subscription/data_source_location.go [hunk 21]
```diff
@@ -0,0 +1,157 @@
...
+			metadata.ResourceData.SetId(fmt.Sprintf("%s/locations/%s", id.ID(), normalizedLocation))
+
+			return metadata.Encode(&state)
+		},
+	}
+}
+
+func getLocation(location string, input *[]resourcesSubscription.Location) (*resourcesSubscription.Location, error) {
+	for _, item := range *input {
+		if pointer.From(item.Name) == location {
```
Comment (catriona-m) [inline internal/services/subscription/data_source_location.go:132]: could we filter the location type to only get the physical location here using `location.Metadata.RegionType`?

File: website/docs/d/location.html.markdown [hunk 22]
```diff
@@ -0,0 +1,50 @@
...
+subcategory: "Base"
+layout: "azurerm"
+page_title: "Azure Resource Manager: azurerm_location"
+description: |-
+   Get information of a specific location.
+---
+
+# Data Source: azurerm_location
+
+Use this data source to access information of a specific location.
```
Comment (catriona-m) [inline website/docs/d/location.html.markdown:11]:
```suggestion
Use this data source to access information of a specific physical location.
```

File: website/docs/d/location.html.markdown [hunk 23]
```diff
@@ -0,0 +1,50 @@
...
+* `physical_zone` - The fully qualified physical zone id of availability zone to which logical zone id is mapped to
+
+
+---
+
+## Timeouts
+
+The `timeouts` block allows you to specify [timeouts](https://www.terraform.io/language/resources/syntax#operation-timeouts) for certain actions:
+
+* `read` - (Defaults to 5 minutes) Used when retrieving the Extended Locations.
```
Comment (catriona-m) [inline website/docs/d/location.html.markdown:50]:
```suggestion
* `read` - (Defaults to 5 minutes) Used when retrieving the Location.
```

**Review by catriona-m — CHANGES_REQUESTED** [review]
Hi @HappyTobi I had a look through this and left a couple of comments mostly around restricting this to retrieving physical locations only since we use the phrase `location` to refer to physical locations in the provider. Once those are fixed up we can take another look. Thanks!

**Comment by HappyTobi** [comment]
@catriona-m  
any updates?

**Review by stephybun — APPROVED** [review]
Thanks @HappyTobi LGTM 👍

**Pull request merged**